### PR TITLE
Add support for EU tax / VAT calculations

### PR DESCRIPTION
**Pull request by mpociot**
## Summary
This PR adds support for EU related tax, VAT and invoice regulations. The goal is to allow EU citizens to use Spark without having to deal with all the regulations and VAT rules.

### Packages
The overall VAT calculation and VAT-ID checking logic is handled in the [VAT-Calculator](https://github.com/mpociot/vat-calculator) package, which is now required in the `composer.json` file.

### Installer
The installer has a `--european-union` flag and will ask the user upon installation if the EU helper should be installed (defaults to false).

When enabling the european flag, the installer will do the following things:

* Publish the VAT-Calculator JS file
* Copy the `EuropeanUnionUser` stub (this includes the `BillableWithinTheEU` trait)
* Copy the `invoice.blade.php` to `resources/vendors/cashier/`
* Add a `SPARK_EUROPEAN` environment variable

### Registration form

To add EU support, this PR will add quite a few form fields to the registration form.

These fields only get added to the registration form, if a *paid* plan is selected!

All of those fields will only be shown, when the `SPARK_EU` variable is set to `true`.

The form uses the VAT-Calculator Javascript library to automatically:

* Prepopulate the country dropdown by trying to detect the users country via IP lookup
* Calculate VAT, Subtotal, Total based on the selected country from the dropdown
* Validate VAT-IDs and update the calculation result accordingly

registration_form

### Registration process

The default `SparkServiceProvider` has a EU related callback, which is used to validate the registration form. 
After the registration succeeds, the Stripe JS API will be used to append the billing address data to stripe. This data will be used on the invoices.

Stripe also get's the tax percentage to calculate all invoices correctly.

stripe_example

### Subscription settings

Similar to the registration process, the billing address fields need to be changeable. Therefor the Subscription tab contains a new panel to update the user's billing address. This is also only visible when the EU setting is turned on.

subscription_settings

### Changing plans

The default `SparkServiceProvider` has a EU related callback, which is used to swap plans. This is needed, because the default `swapAndInvoice` function would invoice the user with 0% tax/vat.

So by using the custom swap callback, we make sure that the Invoice contains the correct tax percentage.

### Invoices

The invoice is modified for usage within the EU. It will show the tax percentage used and will also display a text when a valid VAT-ID was given to show that the VAT was "reverse charged".

invoice_example

### Subscribing from a free plan

If the user registers by selecting a free plan, we need to reproduce the registration process (and all EU related fields) within the settings tab as well.

This form behaves just like the registration form and adds support for VAT calculation and VAT-ID validation.

subscribe_from_free_plan

### CODE REVIEW

**Comment by pavankataria** [comment]
Awesome. This will be very useful to have in Spark +1

**Comment by aaranmcguire** [comment]
Awesome! Great work guys.

**Comment by svenluijten** [comment]
Great! I'd love to see this in Spark. +1

**Comment by Kevinjohn** [comment]
Really like this, and congrats to every involved.
Concerned though that we're using the term "European" (no, I'm not being PC), just that it's not European  countries that this law is catching. It's EU countries (which is a different group from Europeans and the EEA members).

Could we change the code to be Spark::isInEU ? I'm just trying to make it future proof.

**Comment by mpociot** [comment]
@Kevinjohn you're right - `Spark::isInEU` might be the better term for this.
Just changed it.

**Comment by Kevinjohn** [comment]
@mpociot Awesome. Thank you!

**Comment by Keiron-BWS** [comment]
+1 For this, definitely needed - great work!

**Comment by Luceos** [comment]
Amazing Marcel, exactly what I needed!

**Comment by ivuorinen** [comment]
:+1: Very nice work.

**Comment by Keiron-BWS** [comment]
@mpociot Does this have some way of recording the transactions to create a VATMOSS report for your accounts showing the split to different countries or would you expect that to be taken from an export in Stripe?

**Comment by mpociot** [comment]
No it doesn't. I would delegate this to stripe. This also doesn't create sequential invoice numbers, so you might need to take care of this too.

**Comment by kelvinj** [comment]
Good work people.

Has anyone read this [techcrunch article about more countries persuing this model](http://techcrunch.com/2015/10/10/vatmess-is-about-to-get-much-messier/)?

It seems that this model of collecting taxes from overseas is spreading, which could either:

1. widen the scope of this work
2. bring into question where all of this is type of tax stuff is handled - it could easily become to unwieldy

**Comment by Kevinjohn** [comment]
@kelvinj I'd suggest not widening the scope.

The article is jumping the gun a little ( it's actually jumping the shark ). There are already strict laws in the EU  ( and especially in the EEA ) regarding taxation and transferring of money collected between countries; so when the EU consultation began, it only took a mere 8 years for this legislation to come to pass.

The other countries that are 'considering' starting a 'consulting' process (and therefore haven't even decided on consulting on whether they would decide to implement any such idea before attempting to make law) still have a very long way to go.

Who know's, Stripe might even work in those countries before then!

**Comment by kelvinj** [comment]
@Kevinjohn Kevin, thanks for the info & perspective.

**Comment by ronlobo** [comment]
:+1:

File: app/Http/Controllers/Settings/SubscriptionController.php [hunk 1]
```diff
@@ -128,6 +128,43 @@ public function updateCard(Request $request)
...
+            'name' => 'required',
+            'street' => 'required',
+            'zip' => 'required',
+            'city' => 'required',
+            'country' => 'required',
+        ]);
+
+        $customer = Auth::user()->subscription()->getStripeCustomer();
+
+        if ($request->has('company')) {
```
Comment (lucasmichot) [inline app/Http/Controllers/Settings/SubscriptionController.php]: `$request->has('company');` is useless in this case.
`$request->get('company');` or `$request->get('company', null);` are just fine

File: app/Http/Controllers/API/SubscriptionController.php [hunk 2]
```diff
@@ -3,6 +3,7 @@
 namespace Laravel\Spark\Http\Controllers\API;
 
 use Exception;
+use Illuminate\Support\Facades\Auth;
```
Comment (lucasmichot) [inline app/Http/Controllers/API/SubscriptionController.php]: The longest string should be a the the bottom of the imports.

**Comment by mstroeck** [comment]
:+1: 

This is in fact pretty essential - awesome work.

**Comment by mpociot** [comment]
@lucasmichot thank you for looking through the code :)
I applied the changes you mentioned.

**Comment by srmklive** [comment]
@mpociot Awesome work man :). I had to integrate it in my project and believe me you saved me a lot of time coding. Btw do rebase your commits, there are a lot of them :).

**Comment by crash13override** [comment]
+1 for this!!!!
It would be soooo good for EU companies to have it included in Spark. It's really a pain to make all this work right and what you did here is absolutely excellent!

**Comment by kayyyy** [comment]
Awesome :+1: 
Why is this PR not merged yet? :)

**Comment by CoolGoose** [comment]
+1 on this, any way we need to help to get this merged ?

**Comment by domioanna** [comment]
+1 let make this happen

**Comment by srmklive** [comment]
+1 lets make it happen.

**Comment by srmklive** [comment]
@mpociot rebase your commits. There are 54 commits in your pull request. Your pull request should contain only one commit. I think thats why your PR hasn't been merged :

**Comment by taylorotwell** [comment]
Initial thoughts on this. I don't think collecting billing addresses should necessarily be tied to EU / VAT stuff. I think we need to do that as a separate feature that can be used even without EU / VAT stuff. Even for US based SaaS services, collecting billing addresses can reduce the amount of credit charge failures you receive on your app (I know this from Forge, heh). So I think we should tackle the address stuff in Spark first then come back to this EU stuff on top of that.

**Comment by mpociot** [comment]
@taylorotwell Do you want this to happen in a separate PR?
A huge part of this feature is about managing the billing address in Spark.

What I would do right now is:

Create a new branch based on this feature branch, rip out all of the billing address features and let it have it's own configuration to enable / disable the billing address form.

The pure VAT / EU feature would then consist of the VAT calculation, Cashier integration, the invoice template and all the provided settings for validating the registration and swapping plans.

**Comment by taylorotwell** [comment]
Yeah I think we should look at the address stuff in a separate PR totally separate from VAT.

**Comment by Keiron-BWS** [comment]
Getting two pieces of information that agree about a customers location is a large part of the EU VAT issue, so having the billing address is key. So long as the address part goes in at some point I'd be happy but it is needed with this.

**Comment by srmklive** [comment]
@Keiron-BWS i already have a PR #121  in this regard

**Comment by mpociot** [comment]
And maybe you should change the name of the AddressBook model to "BillingAddress".
Because that's what it is. It's not a book, it contains only one address.

**Comment by JimmyRittenborg** [comment]
:+1:

**Comment by xitude** [comment]
Brilliant :+1:

**Comment by taylorotwell** [comment]
Closing this for now but I will be implementing support for this.

**Comment by crash13override** [comment]
@taylorotwell any idea if this will be included "soonish" or later? What I mean is: is it a matter of weeks, months, or a year? I know it's not nice to put pressure on you with all the great things you're already doing, but I'm considering using Spark for a new SaaS project I'm about to launch and it would be super mega helpful to use it and avoiding all the headaches :)
thanks again a lot for all your hard work!

**Comment by taylorotwell** [comment]
Probably a few months

**Comment by KHucks** [comment]
Just as a hint that might be a helpful starting point for someone around the tax handling issues:

Stripe itself suggests to use 3rd Party Services to handle the issues around tax compliance (including VAT MOSS for the EU). As far as I can tell they come at a monthly fee but I decided to give them a try once Spark get's into beta (no experience with them yet)

Stripe specifically names these services: 
For calculating tax rates and ensuring compliance: Taxamo, Avalara
Specifically for invoicing: Quademo, Octobat

Our check out the stripe article yourself - you find it here: https://support.stripe.com/questions/how-does-stripe-help-with-vat-moss

**Comment by mpociot** [comment]
Yeah - the approach I'm taking using my VAT Calculator package is inspired by Quaderno and uses a similar API.

So the main benefit of using one of the 3rd party services vs. my VAT package is that the VAT rates don't wait for pull requests ;)

And most of these services also take care of generating invoices for you - which was also part of this PR.

**Comment by JimmyRittenborg** [comment]
@mpociot is totally right, his package is actually currently the only legit way to go about this stuff without adding a unnecessary cost of a third part service. Impressive work by the way :+1: 

Even if you're a seller of internet services in the US today, you're currently holding the legal sole responsibility for charging the current applicable VAT rate of a private consumer in a EU members state, you're selling to.

His package checks the VAT rate to charge, directly against current official EU sources and also checks if a given VAT number is valid (in which case the VAT can be ignored). It's pretty simple. And is in place to plan out a very uneven competition you may have had against big ass corporations just placing themselves in the member state with the most attractive VAT rate (Luxembourg).

**Pull request closed without merging**